report_dynamic/models/report_dynamic.py

```diff
@@ -0,0 +1,271 @@
+            "res_model": "report.dynamic.section",
+            "view_mode": "tree,form",
+            "target": "current",
+            "context": {"default_report_id": self.id},
+            "domain": [("id", "in", self.section_ids.ids)],
+        }
+
+    def action_wizard_lock_report(self):
+        self.ensure_one()
+        return {
+            "type": "ir.actions.act_window",
+            "res_model": "wizard.lock.report",
+            "view_mode": "form",
+            "target": "new",
+        }
+
+    def action_preview_content(self):
+        self.ensure_one()
+        action = self.env.ref("report_dynamic.report_dynamic_document_preview").read(
+            []
+        )[0]
+        return action
+
+    def action_toggle_active(self):
+        self.ensure_one()
+        self.active = not self.active
+
+    # Override create() and write() to keep
+    # resource_ref always the same with template
+    # even if template.resource_ref=False
+    @api.model
+    def create(self, values):
+        records = super().create(values)
+        for this in records:
+            if this.template_id.resource_ref and not this.res_id:
+                this.resource_ref = this.template_id.resource_ref
+                # Give a default to wrapper_report_id when
+                # user sets template_id
+                this.wrapper_report_id = this._set_wrapper_report_id(this.template_id)
+        return records
+
+    def write(self, values):
+        # If the template is changed back to a non-template
+        # (eg is_template is set to False),
+        # and the template already has children, then disallow.
+        if all(
+            [
+                self.report_ids,
+                self.is_template,
+                "is_template" in values,
+                values.get("is_template") is False,
+            ]
+        ):
+            raise UserError(
+                _(
+                    "You cannot switch this template because "
+                    "it has reports connected to it"
+                )
+            )
+        # If the model is changed while
+        # the template already has children, disallow;
+        if all(
+            [
+                self.report_ids,
+                "model_id" in values,
+                self.model_id != self.env["ir.model"].browse(values.get("model_id")),
+            ]
+        ):
+            raise UserError(_("You cannot change model for this report"))
+        if "template_id" in values and values.get("template_id"):
+            # if in a report we set a template
+            template = self.browse(values.get("template_id"))
+            self.resource_ref = template.resource_ref
+            # Give a default to wrapper_report_id when
+            # user sets template_id
+            self.wrapper_report_id = self._set_wrapper_report_id(template)
+        return super().write(values)
+
+    def unlink(self):
+        for this in self:
+            if not this.is_template:
```

Comment:
you cannot delete this record at all?

Reply:
@ntsirintanis Why couldn't we deleted non-template records again?

Reply:
@ntsirintanis I pushed lazy, untested implementations of all the remaining review comments - could you test each and either fix or ping back on the ones that require more thinking

Reply:
@ntsirintanis Now less lazy and tested fixes. Also some other changes that may not have been needed but I was struggling with the `resource_ref` Reference field, it's a nasty construction.

Reply:
Agreed that resource_ref is an ugly business. Haven't put any thought yet on how to avoid using it. Regarding unlinking here, this unlinks the window action for this template, if the template is the last one in the database for that model. It does not stop using from unlinking a template or non-template record.